File: get_sample_fastq_files.rs
extern crate md5;

use std::collections::HashMap;
use std::env;
use std::fs::{self, File};
use std::io::{Read, BufRead, BufReader, Write};
use std::path::Path;
use md5::{Md5, Digest};
use walkdir::WalkDir;


fn main() {
    let args: Vec<String> = env::args().collect();
    let path_to_search = args.get(1).cloned().unwrap_or_else(|| ".".to_string());
    let sampleid = args.get(2).cloned().unwrap_or_else(|| "".to_string());
    
    let path_sanitized = path_to_search.replace(|c: char| c.is_whitespace() || c == '/' || c == '\\', "_");
    
    let sample_file_path = format!("{}/sample_collection_{}.sample_lines.tsv", path_sanitized, sampleid);
    let files_file_path = format!("{}/sample_collection_{}.files_md5sum_lines.tsv", path_sanitized, sampleid);
    
    let mut data: HashMap<String, HashMap<String, String>> = HashMap::new();
    
    //let re1 = Regex::new(r".*/(.*?)_(S.*L\d{3}).*(R[12]|I1).*\\.fastq\\.gz").unwrap();
    //let re2 = Regex::new(r".*/(.*?)_(S\d+).*(R[12]|I1).*\\.fastq\\.gz").unwrap();
    
    for entry in WalkDir::new(&path_to_search).into_iter().filter_map(Result::ok) {
        let file_path = entry.path();
        if let Some(file_name) = file_path.file_name().and_then(|n| n.to_str()) {
            if file_name.ends_with(".fastq.gz") {
                let path_str = file_path.to_string_lossy().to_string();
                //if let Some(captures) = parse_filename(&path_str, &re1, &re2) {
                if let Some(captures) = parse_filename_split(&path_str) {
                    let (sample, lane, read_type) = captures;
                    data.entry(format!("{}_{}", sample, lane))
                        .or_default()
                        .insert(read_type, path_str);
                }
            }
        }
    }
    
    write_sample_files(&sample_file_path, &data);
    write_md5_files(&files_file_path, &data);
    println!("Data written to '{}'", sample_file_path);
}


fn parse_filename_split(file_path: &str) -> Option<(String, String, String)> {
    let parts: Vec<&str> = file_path.split('/').last()?.split('_').collect();

    if parts.len() < 3 {
        return None;
    }

    // Read type is typically R1, R2, or I1, usually near the end
    let read_type = parts.last()?.to_string();

    // Find lane information, which often has 'S' and 'L' markers
    let lane_idx = parts.iter().position(|&s| s.starts_with('S'))?;
    let lane = parts.get(lane_idx)?.to_string();

    // The sample name is everything before the lane info
    let sample = parts[..lane_idx].join("_");

    Some((sample, lane, read_type))
}

/*fn parse_filename(file_path: &str, re1: &Regex, re2: &Regex) -> Option<(String, String, String)> {
    if let Some(caps) = re1.captures(file_path) {
        return Some((caps[1].to_string(), caps[2].to_string(), caps[3].to_string()));
    }
    if let Some(caps) = re2.captures(file_path) {
        return Some((caps[1].to_string(), caps[2].to_string(), caps[3].to_string()));
    }
    None
}*/


fn write_sample_files(path: &str, data: &HashMap<String, HashMap<String, String>>) {
    let mut file = File::create(path).expect("Could not create sample file");
    writeln!(file, "Sample_Lane\tR1\tR2\tI1").unwrap();
    for (sample_lane, reads) in data {
        writeln!(file, "{}\t{}\t{}\t{}", 
                 sample_lane,
                 reads.get("R1").unwrap_or(&"MISSING_R1".to_string()),
                 reads.get("R2").unwrap_or(&"MISSING_R2".to_string()),
                 reads.get("I1").unwrap_or(&"MISSING_I1".to_string())
        ).unwrap();
    }
}

fn write_md5_files(path: &str, data: &HashMap<String, HashMap<String, String>>) {
    let mut file = File::create(path).expect("Could not create md5 file");
    writeln!(file, "file_name\tmd5sum").unwrap();
    for reads in data.values() {
        for file_path in reads.values() {
            let md5sum = get_md5sum(file_path);
            writeln!(file, "{}\t{}", file_path, md5sum).unwrap();
        }
    }
}

fn get_md5sum(file_path: &str) -> String {
    let path = Path::new(file_path);
    let md5_file = path.with_extension("fastq.gz.md5sum");
    if md5_file.exists() {
        if let Ok(file) = File::open(&md5_file) {
            let reader = BufReader::new(file);
            if let Some(Ok(line)) = reader.lines().next() {
                return line;
            }
        }
    }
    if let Ok(mut file) = File::open(path) {
        let mut hasher = Md5::new();
        let mut buffer = vec![0; 8192];
        while let Ok(bytes) = file.read(&mut buffer) {
            if bytes == 0 { break; }
            hasher.update(&buffer[..bytes]);
        }
        let result = format!("{:x}", hasher.finalize());
        let _ = fs::write(&md5_file, &result);
        return result;
    }
    "none".to_string()
}

File: get_sample_fastq_files.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use Digest::MD5;


# Hash to store organized data
my %data;

# Find and process .fastq.gz files
my $path_to_search = shift || "."; # Default to current directory
my $sampleid = shift || ""; # Default to none ""

my $path = $path_to_search;
$path =~ s/[\/\\\s]+/_/g; 

open my $SAMPLE ,"> $path/sample_collection_$sampleid.sample_lines.tsv" or die "couln not create '$path./sample_collection_$sampleid.sample_lines.tsv'\n";
open my $FILES ,"> $path/sample_collection_$sampleid.files_md5sum_lines.tsv" or die "couln not create '$path./sample_collection_$sampleid.files_md5sum_lines.tsv'\n";

open my $SAMPLE_BN ,"> $path/sample_collection_basename_$sampleid.sample_lines.tsv" or die "couln not create '$path./sample_collection_basename_$sampleid.sample_lines.tsv'\n";
open my $FILES_BN ,"> $path/sample_collection_basename_$sampleid.files_md5sum_lines.tsv" or die "couln not create '$path./sample_collection_basename_$sampleid.files_md5sum_lines.tsv'\n";


open my $find, "-|", "find $path_to_search -name '$sampleid*.fastq.gz'" or die "Failed to execute find: $!\n";

while (<$find>) {
    chomp;
    # Parse filename, e.g., sample_L001_R1_001.fastq.gz
    print( "." );
    #NextSeq7/HVHNYBGXB/outs/fastq_path/Chromium_20191011/SI-GA-B1_3/Sample_16_S7_L001_R1_001.fastq.gz
    #data/files/2025_001/0_fastq/AGCYWLM5/negative_1_S8_R2_001.fastq.gz
    if (m!.*/(.*)_(S.*L\d{3}).*(R[12]|I1).*\.fastq\.gz!) {
        my ($sample, $lane, $read_type) = ($1, $2, $3);
	#print("${sample}_$lane\n");
        $data{"${sample}_$lane"}{$read_type} = $_;
    }elsif (m!.*/(.*)_(S\d+).*(R[12]|I1).*\.fastq\.gz!) {
	my ($sample, $lane, $read_type) = ($1, $2, $3);
	#print("${sample}_$lane\n");
        $data{"${sample}_$lane"}{$read_type} = $_;
    }
    $| = 1;  # Force output flushin
}
close $find;

# Print organized data
print $SAMPLE join("\t", "Sample_Lane", "R1", "R2", "I1"), "\n";
print $SAMPLE_BN join("\t", "Sample_Lane", "R1", "R2", "I1"), "\n";

foreach my $sample_lane (sort keys %data) {
    #for my $file in %{$data{$sample_lane}}{
    #    check_and_add_md5sum( $file );
    #}    
    print $SAMPLE join("\t", 
        $sample_lane,
        $data{$sample_lane}{R1} // "MISSING_R1",
        $data{$sample_lane}{R2} // "MISSING_R2",
        $data{$sample_lane}{I1} // "MISSING_I1"
    ), "\n";
    print $SAMPLE_BN join("\t",
        $sample_lane,
        basename($data{$sample_lane}{R1} // "MISSING_R1"),
        basename($data{$sample_lane}{R2} // "MISSING_R2"),
        basename($data{$sample_lane}{I1} // "MISSING_I1")
    ), "\n";


}


# Print organized data
print $FILES join("\t", "file_name", "md5sum"), "\n";
print $FILES_BN join("\t", "file_name", "md5sum"), "\n";

foreach my $sample_lane (sort keys %data) {
    #for my $file in %{$data{$sample_lane}}{
    #    check_and_add_md5sum( $file );
    #}
    print $FILES join("\n",
        $data{$sample_lane}{R1}."\t".get_md5sum($data{$sample_lane}{R1}) // "MISSING_R1\tnone",
        $data{$sample_lane}{R2}."\t".get_md5sum($data{$sample_lane}{R2}) // "MISSING_R2\tnone",
        $data{$sample_lane}{I1}."\t".get_md5sum($data{$sample_lane}{I1}) // "MISSING_I1\tnone"
    ), "\n";
    print $FILES_BN join("\n",
        basename($data{$sample_lane}{R1}// "MISSING_R1")."\t".get_md5sum($data{$sample_lane}{R1}) // "MISSING_R1\tnone",
        basename($data{$sample_lane}{R2}// "MISSING_R1")."\t".get_md5sum($data{$sample_lane}{R2}) // "MISSING_R2\tnone",
        basename($data{$sample_lane}{I1}// "MISSING_R1")."\t".get_md5sum($data{$sample_lane}{I1}) // "MISSING_I1\tnone"
    ), "\n";

}

close $SAMPLE ;
close $FILES ;

print("Data written to '$path.$sampleid.sample_lines.tsv' and '$path.$sampleid.files_md5sum_lines.tsv'\n");

# Function to check and add md5sum calculation command to the script
sub check_and_add_md5sum {
    my $file = shift;
    
    # Check if the file and its corresponding .md5sum file exist
    my $md5file = $file . ".md5sum";
    
    if (-e $md5file) {
        print STDERR "MD5sum already exists for: $file\n";
    } else {
        # If the .md5sum file doesn't exist, add the command to the calculate.md5sums.sh script
        open my $sh_script, '>>', 'calculate.md5sums.sh' or die "Cannot open calculate.md5sums.sh: $!\n";
        print $sh_script "md5sum $file > $md5file\n";
        close $sh_script;
        print STDERR "Added md5sum calculation command for $file to calculate.md5sums.sh\n";
    }
}

# Function to check if the md5sum file exists and return it, or calculate it if not
sub get_md5sum {
    my $file = shift;

    if (undef $file) {
	return "";
    }
    unless ( -f $file ) {
        return "" ;
    }
    # Create the expected .md5sum file name
    my $md5file = $file . ".md5sum";

    # If the .md5sum file exists, read and return its content
    if (-e $md5file) {
        open my $fh, '<', $md5file or die "Cannot open $md5file: $!\n";
        my $md5sum = <$fh>;
        close $fh;
        chomp $md5sum;  # Remove the newline
        return $md5sum;
    }
    else {
        # If no .md5sum file, calculate and return the MD5 sum
        open my $fh, '<', $file or die "Cannot open $file: $!\n";
        my $md5 = Digest::MD5->new;

        # Read the file and calculate its MD5 sum
        $md5->addfile($fh);
        close $fh;

        # Get the MD5 sum as a hexadecimal string
        my $md5sum = $md5->hexdigest;

        # Optionally, save this MD5 sum to a file
        open my $out_fh, '>', $md5file or die "Cannot write to $md5file: $!\n";
        print $out_fh "$md5sum\n";
        close $out_fh;

        # Return the calculated MD5 sum
        return $md5sum;
    }
}

